### PR TITLE
Use new algorithm for generating patterns from neo skeleta

### PR DESCRIPTION
**Pull request by sffc**
#1317

This mostly resolves pattern generation issues arising from the year/fullyear/era discussion in CLDR Design WG.

The `Year` field should resolve to a narrow (usually 2-digit) year when doing so is customary for the locale/calendar. I determine this by looking at the length pattern. At the same time, the `Month` field should also generally be 2-digit if the year is 2-digit. Same applies to second/minute with hour. And also, the `Day` field should be 2-digit if the month is 2-digit...

That condition was too hard to code manually, so I instead am using a completely new algorithm for generating the patterns. I take the date/time length pattern and simply filter for the fields that are present in the semantic skeleton. This should make the lengths be compatible with each other.

I'm looking for an initial review that verifies that this makes sense with a focus on the JSON files.

### CODE REVIEW

**Comment by eggrobin** [comment]
> I'm looking for an initial review that verifies that this makes sense with a focus on the JSON files.

That is made somewhat difficult by the fact that there appear to be 2904 files changed, so that looking at the diff in any file corresponding to a language I speak is beyond the ability of the GitHub review tool. Spot-checking things in https://github.com/sffc/omnicu/tree/newalg/provider/source/data/debug/datetime/patterns/gregory/skeleton%401 I do not see anything obviously wrong, but frankly I am not sure what to look for either.

As I think I have said in the past, this stuff tends to be easiest to review end-to-end, looking at actual formatted dates rather than collections of patterns.

File: components/datetime/src/neo_skeleton.rs [hunk 1]
```diff
@@ -78,6 +79,26 @@ impl NeoSkeletonLength {
...
+    pub fn to_date_style(self) -> options::length::Date {
+        match self {
+            Self::Long => options::length::Date::Long,
+            Self::Medium => options::length::Date::Medium,
+            Self::Short => options::length::Date::Short,
+        }
+    }
+
+    /// Returns the time style corresponding to this length.
+    pub fn to_time_style(self) -> options::length::Time {
```
Comment (Manishearth) [inline components/datetime/src/neo_skeleton.rs:53]: thought: all of these could be Into impls perhaps (non blocking, not necessary to do at all)

**Review by Manishearth — APPROVED** [review]

**Comment by sffc** [comment]
@eggrobin Do you have any feedback on the mapping algorithm, described in the OP and in the .rs files?

It is a departure from the mapping algorithm I had been using, which came from your original doc.

**Comment by eggrobin** [comment]
>  I take the date/time length pattern and simply filter for the fields that are present in the semantic skeleton.

Do you mean (from a CLDR xml standpoint) filtering the `id` attribute of the `dateFormatItem` element, or do you mean the actual pattern?

**Comment by sffc** [comment]
Maybe best if I illustrate.

Currently algorithm: map semantic skeleta to classical skeleta by inheriting the length directly from the semantic skeleton length according to the table in the doc (short typically means numeric like "M", medium typically means abbreviated like "MMM", long typically means "MMMM").

New algorithm: map semantic skeleta to classical skeleta based on the widths in the dateFormats (or dateSkeletons: they don't seem to be different in practice).

For example, in en Gregorian:

```
            "dateFormats": {
              "full": "EEEE, MMMM d, y",
              "long": "MMMM d, y",
              "medium": "MMM d, y",
              "short": "M/d/yy"
            },
            "dateSkeletons": {
              "full": "yMMMMEEEEd",
              "long": "yMMMMd",
              "medium": "yMMMd",
              "short": "yyMd"
            },
```

The classical skeleton mapping changes as follows:

| Field Set | Semantic Length | Old Classical | New Classical |
|---|---|---|---|
| { Year, Month } | Short | yM | yyM |
| { Year, Month } | Medium | yMMMd | (same) |
| { Year } | Short | y | yy |

**Comment by sffc** [comment]
The effect on formatting is that datagen uses it for selecting the patterns. These classical skeleta only exist very briefly in memory during datagen.

**Comment by sffc** [comment]
ok I think this is ready to land now.

**Comment by eggrobin** [comment]
> These classical skeleta only exist very briefly in memory during datagen.

I understand that; this is why I am asking what the _actual effect_ is, because these endoskeleta are an implementation detail. Can you produce a similar before/after table with patterns, or better, with formatted dates?

**Comment by sffc** [comment]
^ force-push to clean up commit history and make it easier to answer @eggrobin's query

**Comment by sffc** [comment]
Pulling a few examples out of the JSON file changes that I think illustrate the impact:

| Locale | Calendar | Field Set | Old Patterns | New Patterns | Comments |
|---|---|---|---|---|---|
| sr-Latn | Buddhist | { FullYear, Month } | MMMM y. GGGG (long)<br><br>MMM y. GGG (medium, short) | MMMM y. GGG (long)<br><br>MMM y. GGG (medium)<br><br>MMM y. GGGGG (short) | Short date pattern has a GGGGG. None of the patterns use a GGGG (long era). Other locales have similar changes. |
| es | Buddhist | { Year, Month } | MMMM 'de' y G (long)<br><br>MMM y G (medium)<br><br>M/y GGGGG (short) | MMMM 'de' y GGG (long)<br><br>MMM y GGG (medium)<br><br>MMM y GGGGG (short) | The short pattern previously picked up the pattern for the "yM" classical skeleton, which results in "M/y GGGGG". Now it inherits the era from the date pattern and uses the "GGGGGyMMM" classical skeleton, which results in "MMM y GGGGG" |
| ar | Gregorian | { Month } | LLLL (long)<br><br>LLL (medium)<br><br>L (short) | LLLL (long)<br><br>LL (medium)<br><br>L (short) | The medium length changes from LLL (abbreviated) to LL (2-digit) since that is what the lengths are in the date patterns. Might be better the old way? Add this as a special case? (similar changes in other locales) |
| en | Gregorian | { Year } | y (long, medium, short) | y (long, medium)<br><br>yy (short) | This is one of the motivations for the change. People in en-US are weird and apparently like 2-digit years in short mode. This also impacts year-month-day. Classical skeleta are unable to encode the difference between "I want a 4-digit year" and "I want the default short numeric year", which is a separate bug but nonetheless one we need to work around. |
| ja | Gregorian | { Year, Month, Day } | y年M月d日 (long, medium)<br><br>y/M/d (short) | y/M/d (long)<br><br>y/MM/dd (medium, short) | This is a sad consequence of garbage CLDR data. The patterns with Kanji are selected only if the skeleton has MMM, which doesn't happen since the date patterns have an M. This bug is tracked in [CLDR-14993](https://unicode-org.atlassian.net/browse/CLDR-14993). |

Other notes:

- The changes to the `a1e` files are purely the result of a bug fix that this change uncovered.
- The `EraYear` field is adopting the behavior of the new `FullYear` field which has both era-optional and era-required patterns in the same data payload, for locales/calendars that support optional eras.

Spot-checking, I believe that all changes fit into one of the above categories; it's possible I overlooked some corner cases that changed.

**Comment by sffc** [comment]
The above post is limited to changes in the date patterns. I should also document some changes to the time patterns.

| Locale | Field Set | Old Patterns | New Patterns | Comments |
|---|---|---|---|---|
| sr-Latn | { Hour12 } | h a | hh a | I'm still trying to figure out where the length difference comes from. |
| es | { Hour, Minute, Second } | HH:mm:ss | H:mm:ss | The new pattern comes directly from the length patterns. I consider this an improvement. |

**Comment by eggrobin** [comment]
ja | Gregorian | { Year, Month, Day } | y年M月d日 (long, medium)y/M/d (short) | y/M/d (long)y/MM/dd (medium, short) | This is a sad consequence of garbage CLDR data. The patterns with Kanji are selected only if the skeleton has MMM, which doesn't happen since the date patterns have an M. This bug is tracked in CLDR-14993.
-- | -- | -- | -- | -- | --

I do not understand the explanation. The skeleton that gives the y年M月d日 pattern _has_ MMM:
```
						<dateFormatItem id="yMMMd">y年M月d日</dateFormatItem>
```


ar | Gregorian | { Month } | LLLL (long)LLL (medium)L (short) | LLLL (long)LL (medium)L (short) | The medium length changes from LLL (abbreviated) to LL (2-digit) since that is what the lengths are in the date patterns. Might be better the old way? Add this as a special case? (similar changes in other locales)
-- | -- | -- | -- | -- | --

This clearly does not match the intent; a medium month is meant to be something like "Oct.", not "10" (which is a format that is nearly unusable).

**Comment by sffc** [comment]
I pushed fixes for the standalone month and hour widths.

On the Japanese diff: the data is

```
            "dateFormats": {
              "full": "y年M月d日EEEE",
              "long": "y年M月d日",
              "medium": "y/MM/dd",
              "short": "y/MM/dd"
            },
            "dateSkeletons": {
              "full": "yMEEEEd",
              "long": "yMd",
              "medium": "yMMdd",
              "short": "yMMdd"
            },
```

The new algorithm is "pull the month length from dateSkeletons, which in practice is always the same as dateFormats, and use that when running the classical skeleton datetime pattern generation". I can work around this for the { Year, Month, Day } case by just directly using dateFormats in that pattern, instead of running the classical skeleton datetime pattern generation. It doesn't solve it for { Year, Month } or { Month, Day } though.

---

At this point, I believe the behavior in this PR is an overall improvement from the status quo. Therefore, it meets the bar for landing. It is not perfect but it is better than what was there before, so therefore it is landable.

**Comment by eggrobin** [comment]
> The new algorithm is "pull the month length from dateSkeletons, which in practice is always the same as dateFormats, and use that when running the classical skeleton datetime pattern generation".

Ah, I think I get it now. So, for non-Japanese (let’s say French or Italian) medium Gregorian dates with a month, we will get a MMM in the pattern, right?

**Comment by sffc** [comment]
Right:

- French { Year, Month, Day } in Gregorian is unchanged by this PR, aside for the era handling footnote. The fields are the same widths and in the same relative positions.
- Spanish { Year, Month, Day } in Gregorian is unchanged except that the short form uses `yy` instead of `y`.
- South African English { Year, Month, Day } in Gregorian is unchanged except that all 3 forms use `dd` instead of `d`, which comes from the date length pattern.

**Comment by sffc** [comment]
Topic added to the CLDR Design WG for next Monday. We can continue to iterate on this.

**Comment by eggrobin** [comment]
Alright, I think this makes sense then.
If it is not already there (I mostly managed to crash chrome when trying to look too closely at diffs earlier), it might be worth sticking this:

> The new algorithm is "pull the month length from dateSkeletons, which in practice is always the same as dateFormats, and use that when running the classical skeleton datetime pattern generation".

in a comment; it seems like a reasonably understandable spiffy description of the algorithm that I otherwise had not understood until now. Maybe use the actual element name, datetimeSkeleton, and point out that pattern generation is then done based on the patterns over in availableFormats (not those in dateFormats).

**Comment by eggrobin** [comment]
> Topic added to the CLDR Design WG for next Monday. We can continue to iterate on this.

(It is possible I will be on a train, but as I wrote above, this all seems to make sense and to be headed in a reasonable direction, so I am happy leaving it in your capable hands and looking at it asynchronously as needed.)

**Comment by sffc** [comment]
I pushed an extra comment to the branch and added additional documentation in https://docs.google.com/document/d/1dmMk_XODm3DGe84GmMVw7O6a7oIs5yojDXm_25CcMbw/edit#heading=h.a2hszeb6w4ag

**Review by Manishearth — APPROVED** [review]

**Pull request merged**